### PR TITLE
Fsm

### PR DESCRIPTION
**Pull request by rylev**

### CODE REVIEW

**Comment by ghost** [comment]
I want to squash this before we merge, after the other features are in place.

File: README.md [hunk 1]
```diff
@@ -23,57 +23,61 @@ In the future, the structure of the value's contents may change to allow for spe
 
 ## Bootstrapping
 
-On boot Operator relies on an operator.json file to specify configuration of the node as well as the "global" containers that should always be running on the node.
+On boot Operator relies on an operator.json file to specify configuration of the node as well as the "global" containers that should always be running on the node. Any cli flag can also be specified in this json file and will be merged into the already passed cli values.
```
Comment (nathan-ms) [inline README.md]: This is still optimistic. Adding to the wunderlist.

File: boot/commands.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+	"github.com/wakeful-deployment/operator/node"
+	"time"
+)
+
+func DetectOrBootConsul(state *State) error {
+	// TODO
+	return nil
+}
+
+func LoadStateFromFile(path string) *State {
```
Comment (ghost) [inline boot/commands.go]: I'd like to rename this to `LoadBootStateFromFile`

File: boot/commands.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+		return nil
+	}
+
+	return state
+}
+
+func Boot(state *State) {
+	var err error
+
+	global.Transition(global.NewState(global.Booting, nil))
```
Comment (ghost) [inline boot/commands.go]: I want to make a `TransitionState(s State, e error)` that will call `NewState` internally.

File: boot/commands.go [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+	if err != nil {
+		global.Transition(global.NewState(global.NormalizingFailed, err))
+		return
+	}
+
+	global.Transition(global.NewState(global.Booted, nil))
+}
+
+func Run(currentState *State, wait string) {
+	if global.CurrentState.NotEqual(global.Running) && global.CurrentState.NotEqual(global.Booted) {
```
Comment (ghost) [inline boot/commands.go]: `global.CurrentState` and `currentState` are confusing. Let's rename one of them.

File: consul/commands.go [hunk 5]
```diff
@@ -0,0 +1,110 @@
+package consul
```
Comment (ghost) [inline consul/commands.go]: Let's just rename this file to `consul.go`?

File: boot/commands.go [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+		global.Transition(global.NewState(global.NormalizingFailed, err))
+		return
+	}
+
+	global.Transition(global.NewState(global.Running, nil))
+
+	directoryStateUrl.Index = directoryState.Index // for next iteration
+}
+
+func Once(currentState *State) {
```
Comment (ghost) [inline boot/commands.go]: `Once` and `Loop` are now small enough to go back to the main package (or somewhere else).

File: boot/normalize.go [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+func Normalize(desiredState *State, currentNodeState *node.State) error {
+	// always try to fix the containers before fixing the registrations
+
+	var desiredContainers []container.Container
+
+	for _, service := range desiredState.Services {
+		desiredContainers = append(desiredContainers, service.Container())
+	}
+
+	err := docker.NormalizeDockerContainers(desiredContainers, currentNodeState.Containers)
```
Comment (ghost) [inline boot/normalize.go]: Rename to `NormalizeContainers`

File: directory/kv.go [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+	"encoding/json"
+	"github.com/wakeful-deployment/operator/service"
+	"strings"
+)
+
+type KV struct {
+	Key         string
+	Value       string
+	ModifyIndex int
+	Service     service.Service
```
Comment (ghost) [inline directory/kv.go]: Let's remove this property and just have `DecodeService` return it. I don't think we have a need to cache this anymore.

File: docker/commands.go [hunk 9]
```diff
@@ -0,0 +1,137 @@
+package docker
```
Comment (ghost) [inline docker/commands.go]: Maybe also just rename this one to `docker.go`

File: fsm/fsm.go [hunk 10]
```diff
@@ -0,0 +1,46 @@
+package fsm
+
+type State int
```
Comment (ghost) [inline fsm/fsm.go]: Let's migrate our `State struct` from `global` over to here so we can have nicely named states.

File: operator.json [hunk 11]
```diff
@@ -0,0 +1,48 @@
+{
```
Comment (nathan-ms) [inline operator.json]: Need to update this json file with the new format.

**Comment by rylev** [comment]
I think this thing is ready to go!

**Pull request merged**